phpstan.neon.dist

```diff
@@ -19,7 +19,7 @@ parameters:
         - '#Constant TL_SCRIPT not found\.#'
 
         # Ignore the Adapter::missingMethod() warning
-        - '#Call to an undefined method Contao\\CoreBundle\\Framework\\Adapter::missingMethod\(\)\.#'
+        - '#Call to an undefined method Contao\\CoreBundle\\Framework\\Adapter::.+\(\)\.#'
```

Comment:
Why did you change this?

Reply:
I use the adapter to decorate the `Contao\Image::getHtml` method, see https://github.com/contao/contao/pull/1382/files#diff-5ea56ee23b04ee35ebcf7c9d0e8b3df0R63

I got an undefined method warning. I adjusted it so that all missing method warnings would be silenced.

Reply:
But that is not the correct fix, I'm afraid. Instead, you should add an inline phpDoc comment:

```php
        /** Image $adapter */
        $adapter = $this->framework->getAdapter(Image::class);

        return $adapter->getHtml('…');
```

Reply:
Imo the doc comments should also include the adapter:
```php
 /** @var Image&Adapter $adapter */`
```

Reply:
No, that is wrong. The object is not an actual instance of the `Image` class, therefore – if you really want to add the `Adapter` class – it would be `Image|Adapter`. We are not adding the `Adapter` class though.

Reply:
:thinking: Imo omitting `Adapter` will potentially induce problems: You're relying on the static analysis tools to ignore the return type of `$framework->getAdapter()` in this case. I'm not sure if this would even work with higher phpstan levels?

Widening the type is imo the correct approach here (in the sense of duck typing like with a mock object). It won't ever be an `Image` (therefore `Image|Adapter` is wrong)  but it will have all it's properties.

Also, have a look at these examples:
```php
// this will break (invalid return type)
function getImageAdapter(): Image {
    return $this->framework->getAdapter(Image::class);
}

// this won't break but isn't helpful for the consumer (missing type information)
function getImageAdapter(): Adapter {
    return $this->framework->getAdapter(Image::class);
}

// this is typesafe and helpful :-)
/**
* @return Adapter&Image
*/
function getImageAdapter(): Adapter {
    return $this->framework->getAdapter(Image::class);
}
```

Reply:
Using an intersection type is still wrong. 🤷‍♂️ 

> `Foo&Bar` means that the type of a variable is Foo and Bar at the same time.

This condition is clearly not met here!

Reply:
Neighter `\Image` nor `\Image&Adapter` is correct here imho. Using phpstan @mixin should be the way to go. I update the PR accordingly. https://phpstan.org/writing-php-code/phpdocs-basics#mixins

Example:

```php
<?php declare(strict_types = 1);

class HelloWorld
{
	public static function sayHello(DateTimeImmutable $date): void
	{
		echo 'Hello, ' . $date->format('j. n. Y');
	}
}

/** 
 * @template T 
 * @mixin T
 */
class Shouter
{
	/** 
	 * @var class-string<T> 
	 */
	private $className;
	
	/** 
	 * @param class-string<T> $className
	 */
	public function __construct(string $className)
	{
		$this->className = $className;
	}
	
	/** 
	 * @param mixed[] $arguments
	 * @return mixed 
	 */
	public function __calc(string $method, array $arguments)
	{
		return $this->className::$method(... $arguments);
	}
}

$shouter = new Shouter(HelloWorld::class);
$shouter->sayHello(new DateTimeImmutable());
```

Reply:
Resolved by annotating it with the `@mixin` annotation which works pretty well. See ef2cf67.

Reply:
>> Foo&Bar means that the type of a variable is Foo and Bar at the same time.
>
>This condition is clearly not met here!

Of course, it is? It's duck typing. Otherwise you could only ever use Interfaces in intersection types.

Reply:
@dmolineus You are absolutely right. I have encapsulated your changes into a separate PR for Contao 4.9 in #1836.